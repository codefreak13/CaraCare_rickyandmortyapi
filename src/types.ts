interface Episode {
  name: string;
}

export interface Character {
  id: string;
  name: string;
  status: string;
  species: string;
  gender: string;
  origin: {
    name: string;
  };
  episode: Episode[];
  created: string;
  image: string;
}

interface FilterCharacter {
  status?: string;
}

interface Info {
  count: number;
  pages: number;
  next: number;
  prev: number;
}

export interface CharacterVars {
  filter?: FilterCharacter;
  page?: number;
}

export interface CharacterData {
  characters: {
    info: Info;
    results: Character[];
  };
}

export const API = 'https://rickandmortyapi.com/graphql';

export enum FLEX_TYPE {
  columnReverse = 'column-reverse',
  row = 'row',
}

export enum WRAP {
  wrap = 'wrap',
  nowrap = 'nowrap',
  wrapReverse = 'wrap-reverse',
}

export enum ALIGN_TYPE {
  center = 'center',
  flexStart = 'flex-start',
}

export enum VIEW_TYPE {
  list = 'list',
  grid = 'grid',
}

export enum FAV_ICON {
  heart = 'heart',
  heartOutline = 'heart-outline',
}

export enum POSITION_TYPE {
  absolute = 'absolute',
  relative = 'relative',
}

export enum SIZE {
  small = 'small',
  large = 'large',
}

export type ListItemProps = {
  viewType?: string;
  onPress?: () => void;
  addToFavorite?: () => void;
} & Character;

export const enum STORAGE_VALUES {
  FAVORITE_LIST = 'FAVORITE_LIST',
  VIEW_TYPE = 'VIEW_TYPE',
}
