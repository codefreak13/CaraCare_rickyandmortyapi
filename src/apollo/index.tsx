import {ApolloClient, HttpLink, InMemoryCache} from '@apollo/client';

export const client = new ApolloClient({
  link: new HttpLink({uri: 'http://rickandmortyapi.com/graphql'}),
  //   uri: 'http://rickandmortyapi.com/graphql',
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            keyArgs: false,
            merge(existing, incoming) {
              if (!existing) return incoming;
              const newValue = {...existing};
              newValue.info = incoming.info;
              newValue.results = [...existing.results, ...incoming.results];
              return newValue;
            },
          },
        },
      },
    },
  }),
});
