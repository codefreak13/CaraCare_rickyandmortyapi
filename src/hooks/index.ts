export {default as useList} from './useList';
export {default as useListItem} from './useListItem';
export {default as useDetail} from './useDetail';
export {default as useFavorite} from './useFavorite';
export {default as useContextHook} from './useContextHook';
