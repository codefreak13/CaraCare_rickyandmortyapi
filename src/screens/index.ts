export {default as ListScreen} from './ListScreen';
export {default as FavoriteScreen} from './FavoriteScreen';
export {default as DetailScreen} from './DetailScreen';
