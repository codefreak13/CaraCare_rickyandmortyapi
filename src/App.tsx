import React from 'react';
import {client} from './apollo';
import {ApolloProvider} from '@apollo/client';
import AppContextProvider from './store/Context';
import Screens from './navigation/RootNavigator';
import {SafeAreaView, StatusBar, StyleSheet} from 'react-native';

const App = () => {
  return (
    <SafeAreaView style={styles.backgroundStyle}>
      <StatusBar barStyle={'light-content'} />
      <ApolloProvider client={client}>
        <AppContextProvider>
          <Screens />
        </AppContextProvider>
      </ApolloProvider>
    </SafeAreaView>
  );
};

export default App;

const styles = StyleSheet.create({
  backgroundStyle: {
    flex: 1,
  },
});
