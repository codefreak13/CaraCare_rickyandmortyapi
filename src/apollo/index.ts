import {ApolloClient, HttpLink, InMemoryCache} from '@apollo/client';
import {API} from 'src/types';

export const client = new ApolloClient({
  link: new HttpLink({uri: API}),
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            // Don't cache separate results based on
            // any of this field's arguments.
            keyArgs: false,
            // Concatenate the incoming list items with
            // the existing list items.
            merge(existing, incoming) {
              if (!existing) return incoming;
              const newValue = {...existing};
              newValue.info = incoming.info;
              newValue.results = [...existing.results, ...incoming.results];
              return newValue;
            },
          },
        },
      },
    },
  }),
});
