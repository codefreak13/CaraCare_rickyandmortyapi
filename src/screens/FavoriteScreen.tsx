import React from 'react';
import {Header} from 'src/components/ui';
import {FavoriteList} from 'src/components/list-views';
import {useFavorite} from 'src/hooks';

const FavoriteScreen = () => {
  const {favoriteList, goBack, onPress} = useFavorite();

  return (
    <>
      <Header
        leftButton={{
          onclick: () => goBack(),
        }}
        customMiddleIcon
        title="FAVORITE CHARACTERS"
      />
      <FavoriteList data={favoriteList} onPress={onPress} />
    </>
  );
};

export default FavoriteScreen;
