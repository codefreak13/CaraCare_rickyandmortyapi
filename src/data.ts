import {GET_CHARACTERS} from './apollo/gql';

const statusData = [
  {label: 'Alive', value: 'Alive'},
  {label: 'Dead', value: 'Dead'},
];

const mockData = [
  {
    request: {
      query: GET_CHARACTERS,
    },
    results: {
      data: {
        result: [
          {
            id: '123',
            name: 'Mark',
            status: 'Alive',
            gender: 'Male',
            origin: {id: '11', name: 'Earth'},
            species: 'human',
            episode: {id: '11', name: 'Pilot'},
          },
          {
            id: '124',
            name: 'Mary',
            status: 'Dead',
            gender: 'Female',
            origin: {id: '12', name: 'Mars'},
            species: 'human',
            episode: {id: '11', name: 'Final'},
          },
          {
            id: '125',
            name: 'Paul',
            status: 'Unknown',
            gender: 'Male',
            origin: {id: '11', name: 'Earth'},
            species: 'unknnown',
            episode: {id: '11', name: 'Pilot'},
          },
        ],
      },
    },
  },
];

export {statusData, mockData};
