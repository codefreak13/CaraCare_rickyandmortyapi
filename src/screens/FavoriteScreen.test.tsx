import React from 'react';
import {mockData} from 'src/data';
import FavoriteScreen from './FavoriteScreen';
import AppContextProvider from 'src/store/Context';
import {render} from '@testing-library/react-native';
import {MockedProvider} from '@apollo/client/testing';
import {NavigationContainer} from '@react-navigation/native';

describe('FavoriteScreen', () => {
  it('renders FavoriteScreen correctly', () => {
    const {toJSON, getByText} = render(
      <MockedProvider mocks={mockData}>
        <AppContextProvider>
          <NavigationContainer>
            <FavoriteScreen />
          </NavigationContainer>
        </AppContextProvider>
      </MockedProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
    expect(getByText('FAVORITE CHARACTERS')).toBeTruthy();
  });
});
