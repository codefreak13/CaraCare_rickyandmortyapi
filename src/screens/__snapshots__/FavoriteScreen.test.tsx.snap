// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteScreen renders FavoriteScreen correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 83.375,
          "width": 750,
        },
        null,
        null,
      ]
    }
  >
    <View
      style={
        Object {
          "left": 15,
          "position": "absolute",
          "zIndex": 1,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#1D1F20",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome5Free-Solid",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {
                "fontWeight": "900",
              },
            ]
          }
        >
          ÔÅ†
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": undefined,
              "fontSize": 26.5,
              "fontWeight": "bold",
            },
            Object {
              "color": "#1D1F20",
            },
          ]
        }
      >
        FAVORITE CHARACTERS
      </Text>
    </View>
  </View>,
  <RCTScrollView
    ListEmptyComponent={
      <View
        style={
          Object {
            "alignItems": "center",
            "marginTop": 328.5,
          }
        }
      >
        <Unknown
          title="No Item"
        />
      </View>
    }
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "marginTop": 328.5,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": undefined,
                "fontSize": 26.5,
                "fontWeight": "bold",
              },
              undefined,
            ]
          }
        >
          No Item
        </Text>
      </View>
    </View>
  </RCTScrollView>,
]
`;
